Views Tests

damonkelley opened the pull request:
Fixes #17

damonkelley commented:
ping @ldko @somexpert 

This is a big one.

somexpert commented on tests/settings/common.py:
```diff
@@ -13,6 +13,7 @@
     'django.contrib.staticfiles',
     'django.contrib.humanize',
     'premis_event_service',
+    'tests',
```
I'm not sure why this is here. Doesn't seem to be the norm for us, at least not that I've noticed. Perhaps you did something different in this?

damonkelley replied:
Yeah, we have it in some apps. Others we don't.

somexpert commented on tests/test_views.py:
```diff
@@ -0,0 +1,750 @@
+    assert response.status_code == 200
+
+
+def test_json_event_list_returns_ok(rf):
```
What do you think about renaming this and the next couple tests to `test_recent_event_list_*` rather than `test_json_event_list_*`?

damonkelley replied:
I most certainly agree!

somexpert commented:
Do you have any tests for [humanAgent()](https://github.com/unt-libraries/django-premis-event-service/blob/views-tests/premis_event_service/views.py#L388)?

damonkelley commented:
I think I overlooked `humanAgent()`

somexpert commented on tests/test_views.py:
```diff
@@ -0,0 +1,750 @@
+    request = rf.get('/')
+    response = views.json_agent(request, agent.agent_identifier)
+    assert response.status_code == 200
+    assert response.get('Content-Type') == 'application/json'
```
Having both these assertions in the same test seems to diverge from what your first tests looked like. There they were split into separate tests. I don't know that either needs to be changed, but it's something to be aware of.

somexpert commented:
Are you missing a `test_json_agent_raises_404()` test? I ask because you do include such a test in one of the other views that can return a 404.

somexpert commented on tests/test_views.py:
```diff
@@ -0,0 +1,750 @@
+        views.app_agent(request)
+
+    @pytest.fixture
+    def app_agent_xml(self, ):
```
Why the trailing comma?

damonkelley commented on tests/test_views.py:
```diff
@@ -0,0 +1,750 @@
+        response = views.app_agent(request)
+        assert response.status_code == 400
+
+    def test_get_with_invalid_identifier_returns_not_found(self, rf):
```
Change this to `test_invalid_identifier_returns_not_found()`

somexpert commented on tests/test_views.py:
```diff
@@ -0,0 +1,756 @@
+        response = views.app_event(request, event.event_identifier)
+        assert response.status_code == 200
+
+    def test_get_with_identifier_returns_not_found(self, rf):
```
You could probably join this and the `test_delete_returns_not_found` tests into a more generic test like in the previous test class.

damonkelley replied:
This one is a little different. In the `app_agent` view, we were getting the object first, then checking the request methods. In `app_event`, the request method is checked, then the object is fetched. Thus, in `app_event` we are allowing the `Http404` exception to be raised in multiple places.

I will say this though. I think the names need to be modified.
- `test_get_with_invalid_identifier_returns_not_found`
- `test_delete_with_invalid_identifier_returns_not_found`

somexpert replied:
Yep, I see what you mean. Think I was getting confused when skimming back to source and started looking at `app_agent` instead of `app_event`.

damonkelley replied:
No worries, it is easy to get confused about. There is a lot of code there.

somexpert commented:
Seems like we should have a `test_returns_bad_request` in the `TestAppEvent` class similar to the one in the [`TestAppAgent`](https://github.com/unt-libraries/django-premis-event-service/pull/16/files#diff-a4f5c4d64d4811b15e1405573c00f3f4R257) class.

somexpert commented:
In `TestEventSearch`, I don't see any test that checks the number of results per page, while other test classes do check that. Should we add one to this class?

damonkelley commented:
>Are you missing a test_json_agent_raises_404() test? I ask because you do include such a test in one of the other views that can return a 404.

Added in 1b1815f

> In TestEventSearch, I don't see any test that checks the number of results per page, while other test classes do check that. Should we add one to this class?

Added in e275eba

>Do you have any tests for humanAgent()?

Added in d5a4877

somexpert commented:
:+1:

ldko commented on tests/test_views.py:
```diff
@@ -0,0 +1,801 @@
+    assert response.status_code == 200
+
+
+def test_humanAgent_returns_ok(rf):
```
test_humanAgent_returns_ok test is same as test_humanEvent_returns_ok test.

damonkelley replied:
Fixed in fb355eb

ldko commented on tests/test_views.py:
```diff
@@ -0,0 +1,801 @@
+
+
+def test_json_agent_payload(rf):
+    """Verify that the json output has matches the Agent attributes."""
```
docstring may need fixing.

damonkelley commented:
@ldko did you have any other comments?

ldko commented on tests/test_views.py:
```diff
@@ -0,0 +1,801 @@
+        views.app_agent(request)
+
+    def test_returns_bad_request(self, rf):
+        """Check that a request that is not GET, POST, PUT, or DELETE will
```
The intent of the
```
    else:
        return HttpResponseBadRequest("Invalid method for this URL.")
```
seems more about checking to see that if there is not an identifier, you should only be doing a GET or a POST, not a PUT or a DELETE. We may want to open an issue that we should allow HEAD requests. The test to me makes more sense if it is testing that a PUT or DELETE  request without an identifier returns a 400.

ldko commented on tests/test_views.py:
```diff
@@ -0,0 +1,801 @@
+        request = rf.get('?type={0}'.format(event_type))
+        response = views.app_event(request)
+        assert self.event_in_filtered_results(response, event)
+
```
There is not a test here for a GET request with no identifier related to using the `orderby` parameter. I opened up an issue related to a case where it will throw a 500 error. https://github.com/unt-libraries/django-premis-event-service/issues/24.

ldko commented on tests/test_views.py:
```diff
@@ -0,0 +1,801 @@
+    REL_PREVIOUS = 'previous'
+
+    def response_has_entry(self, response, event):
+        """True event is the only Event in the response content."""
```
True if the...

ldko commented on tests/test_views.py:
```diff
@@ -0,0 +1,801 @@
+    RESULTS_PER_PAGE = 20
+
+    def response_has_event(self, response, event):
+        """True event is the only Event in the response context."""
```
True if the...

ldko commented:
@damonkelley ok, I am done commenting.

damonkelley commented:
@ldko, made those changes

ldko commented:
:+1:
